### DISCUSSION TITLE
Does binding/MVVM work with Avalonia AutoCompleteBox?

### DISCUSSION THREAD

**Discussion by anovik** [AvaloniaUI/Avalonia#15104] [General]
Trying to use binding with AutoCompleteBox control but the control is empty in the result.

```
<AutoCompleteBox Text="{Binding CurrentDirectory}" FilterMode="Contains" 
				 ItemsSource="{Binding LatestDirectories}">      
</AutoCompleteBox>
```

The similar properties work with TextBox, for instance, so is there any difference specific for AutoCompleteBox?

**Comment by timunie** [comment]
Should work. Show us your VM property definitions pls.

**Reply by anovik** [reply]
Thanks for your reply, VM properties are not really interesting, just string/ ObservableCollection<string> with getters and setters. 

I did more debugging and found out that binding actually works but after CurrentDirectory is updated with correct values, empty value arrives with the following stack trace (that's why the control is empty in the result). Do you know how to solve this, is it related to one-way/two-way binding? 	

> 
> Avalonia.Base.dll!Avalonia.Data.Core.Plugins.InpcPropertyAccessorPlugin.Accessor.SetValue(object value, Avalonia.Data.BindingPriority priority) Line 55	C#
>  	Avalonia.Base.dll!Avalonia.Data.Core.Plugins.ExceptionValidationPlugin.Validator.SetValue(object value, Avalonia.Data.BindingPriority priority) Line 20	C#
>  	Avalonia.Base.dll!Avalonia.Data.Core.PropertyAccessorNode.SetTargetValueCore(object value, Avalonia.Data.BindingPriority priority) Line 41	C#
>  	Avalonia.Base.dll!Avalonia.Data.Core.BindingExpression.OnNext(object value) Line 162	C#
>  	Avalonia.Base.dll!Avalonia.Reactive.LightweightObservableBase<object>.PublishNext(object value) Line 140	C#
>  	Avalonia.Base.dll!Avalonia.AvaloniaObject.RaisePropertyChanged<string>(Avalonia.AvaloniaProperty<string> property, Avalonia.Data.Optional<string> oldValue, Avalonia.Data.BindingValue<string> newValue, Avalonia.Data.BindingPriority priority, bool isEffectiveValue) Line 421	C#
>  	Avalonia.Base.dll!Avalonia.PropertyStore.EffectiveValue<string>.SetAndRaiseCore(Avalonia.PropertyStore.ValueStore owner, Avalonia.StyledProperty<string> property, string value, Avalonia.Data.BindingPriority priority, bool isOverriddenCurrentValue, bool isCoercedDefaultValue) Line 332	C#
>  	Avalonia.Base.dll!Avalonia.PropertyStore.ValueStore.SetCurrentValue<string>(Avalonia.StyledProperty<string> property, string value) Line 240	C#
>  	Avalonia.Controls.dll!Avalonia.Controls.AutoCompleteBox.UpdateTextValue(string value, bool? userInitiated) Line 1256	C#
>  	Avalonia.Controls.dll!Avalonia.Controls.AutoCompleteBox.TextUpdated(string newText, bool userInitiated) Line 1291	C#
>  	Avalonia.Base.dll!Avalonia.Threading.DispatcherOperation.InvokeCore() Line 250	C#
>  	Avalonia.Base.dll!Avalonia.Threading.DispatcherOperation.Execute() Line 237	C#
>  	Avalonia.Base.dll!Avalonia.Threading.Dispatcher.ExecuteJob(Avalonia.Threading.DispatcherOperation job) Line 692	C#
>  	Avalonia.Base.dll!Avalonia.Threading.Dispatcher.ExecuteJobsCore(bool fromExplicitBackgroundProcessingCallback) Line 720	C#
>  	Avalonia.Base.dll!Avalonia.Threading.Dispatcher.Signaled() Line 702	C#

**Reply by timunie** [reply]
> VM properties are not really interesting, just string/ ObservableCollection with getters and setters.

I'm really interested in those. If they are not defined correctly, that could be the root cause here. For example the ObservableCollection shouldn't have  a setter imo and the string needs to raise PropertyChanged.

**Reply by anovik** [reply]
Yes, that's what I am doing. OK, binding is working, probably this empty string coming is due to some problem in my code. Thanks for your help, will close this.

**Reply by anovik** [reply]
Sorry, one more question. What is the correct way of setting SelectedItem for AutoCompleteBox? 

If I set it similar to Text then the box is populated but I can't type arbitrary text. If it is not set, I can type the text but initial problem happens with empty result at start.

**Reply by timunie** [reply]
I'd try `BindingMode=OneWayToSource` but haven't tried on my own.
